src/logic/graphMapping.ts

```diff
@@ -1,36 +1,121 @@
 import {
-  MappedEdge, 
+  MappedEdge,
   MappedNode,
   Edge,
   MappedGraph,
   GraphIndex,
   DisassembledGraph,
+  Position,
 } from './commonInterfaces';
+const rand = require('random-seed');
+
+const SWITCH_ALGORITHM_ABOVE_X_NODES = 10;
+const CIRCLE_RADIUS = 40;
 
 
 export default function mapGraph(graph: DisassembledGraph): MappedGraph {
   return calcNodes(graph[GraphIndex.NODES], graph[GraphIndex.EDGES]);
 }
 
 function calcNodes(nodes: Array<number>, edges: Array<Edge>):
-  [Array<MappedNode>, Array<MappedEdge>]
-{
-  // TODO better node placing
-  const mappedNodes = nodes.map((l, i) => ({
-    id: i,
-    x: Math.random() * 100,
-    y: Math.random() * 100,
-  }));
+  [Array<MappedNode>, Array<MappedEdge>] {
+
+  const mappedNodes = placeNodes(nodes, edges);
   const mappedEdges = edges.map((edge: Edge, i) => {
     const defaultNode = { x: 0, y: 0 };
-    const node1 = mappedNodes.find((e) => e.id === edge.from) ?? defaultNode;
-    const node2 = mappedNodes.find((e) => e.id === edge.to) ?? defaultNode;
+    const nodeFrom = mappedNodes.find((_, i) => i === edge.from) ?? defaultNode;
+    const nodeTo = mappedNodes.find((_, i) => i === edge.to) ?? defaultNode;
     return {
-      id: i,
       label: edge.label,
-      from: { x: node1.x, y: node1.y },
-      to: { x: node2.x, y: node2.y },
+      from: { x: nodeFrom.x, y: nodeFrom.y },
+      to: { x: nodeTo.x, y: nodeTo.y },
     };
   });
   return [mappedNodes, mappedEdges];
 }
+
+function placeNodes(nodes: Array<number>, edges: Array<Edge>): Array<MappedNode> {
+  if (nodes.length > SWITCH_ALGORITHM_ABOVE_X_NODES) {
+    return placeRandomly(nodes, edges);
+  }
+  return placeOnCircle(nodes, edges);
+}
+
+function placeOnCircle(nodes: Array<number>, edges: Array<Edge>): Array<MappedNode> {
+  const angleStep = 360 / nodes.length;
+  const center = { x: 50, y: 50 };
+  return nodes.map((node: number) =>
+    getPositionOnCircle(node * angleStep, CIRCLE_RADIUS, center));
+}
+
+/**
+ * It won't place nodes too close together. RNG's seed is graph for consistency.
+ */
+function placeRandomly(nodes: Array<number>, edges: Array<Edge>): Array<MappedNode> {
+  const seed = nodes.toString() + edges.toString();
+  const gen = typedGenerator(seed);
+  const dissallowedRadius = getDissallowedRadiusBasedOnDensity(nodes.length);
+  const positions = positionGenerator(gen, dissallowedRadius);
+
+  return nodes.map(_ => positions.next().value);
```

Comment:
All the logic for placing randomly / on circle should be grouped in a class or a separate module